### PR TITLE
adding load native LightGBM model from string and file API

### PR DESCRIPTION
**Pull request by imatiach-msft**
adding load native LightGBM model from string and file API
Fixes issue https://github.com/Azure/mmlspark/issues/293

### CODE REVIEW

File: src/lightgbm/src/main/python/LightGBMClassifier.py [hunk 1]
```diff
@@ -26,13 +26,32 @@ def saveNativeModel(self, sparkSession, filename):
...
+        jsession = sparkSession._jsparkSession
+        return self._java_obj.loadNativeModelFromFile(jsession, filename, labelColName,
+                                                      featuresColName, predictionColName,
+                                                      probColName, rawPredictionColName)
+
+    def loadNativeModelFromString(self, model, labelColName="label",
+                                  featuresColName="features", predictionColName="prediction",
+                                  probColName="probability", rawPredictionColName="rawPrediction"):
+        """
+        Load the model from a native LightGBM text file.
```
[the remark this answers is not in the document]
Comment (imatiach-msft) [reply src/lightgbm/src/main/python/LightGBMClassifier.py]: done

File: src/lightgbm/src/main/python/LightGBMClassifier.py [hunk 2]
```diff
@@ -26,13 +26,32 @@ def saveNativeModel(self, sparkSession, filename):
...
         self._java_obj.saveNativeModel(jsession, filename)
 
+    def loadNativeModelFromFile(self, sparkSession, filename, labelColName="label",
+                                featuresColName="features", predictionColName="prediction",
+                                probColName="probability", rawPredictionColName="rawPrediction"):
+        """
+        Load the model from a native LightGBM text file.
+        """
+        jsession = sparkSession._jsparkSession
+        return self._java_obj.loadNativeModelFromFile(jsession, filename, labelColName,
```
[the remark this answers is not in the document]
Comment (imatiach-msft) [reply src/lightgbm/src/main/python/LightGBMClassifier.py]: done! added an example to notebook :)

File: src/lightgbm/src/test/scala/VerifyLightGBMClassifier.scala [hunk 3]
```diff
@@ -212,6 +212,16 @@ class VerifyLightGBMClassifier extends Benchmarks with EstimatorFuzzing[LightGBM
...
         .fit(testData)
+      // Verify can load model from file
+      val newModelFromString = LightGBMClassificationModel.loadNativeModelFromString(oldModelString,
+        labelColumnName, featuresColumn, rawPredictionColName = rawPredCol)
+      val newModelFromFile = LightGBMClassificationModel.loadNativeModelFromFile(session, modelPath,
+        labelColumnName, featuresColumn, rawPredictionColName = rawPredCol)
+      val originalModelResult = model.transform(testData)
+      val loadedModelFromStringResult = newModelFromString.transform(testData)
+      val loadedModelFromFileResult = newModelFromFile.transform(testData)
+      assert(baseDfEq.areEqual(loadedModelFromStringResult, originalModelResult))
```
Comment (mhamilton723) [inline src/lightgbm/src/test/scala/VerifyLightGBMClassifier.scala]: you can just use === here

File: src/lightgbm/src/main/scala/LightGBMRegressor.scala [hunk 4]
```diff
@@ -120,4 +123,22 @@ class LightGBMRegressionModel(override val uid: String, model: LightGBMBooster,
   def getModel: LightGBMBooster = this.model
 }
 
-object LightGBMRegressionModel extends ConstructorReadable[LightGBMRegressionModel]
+object LightGBMRegressionModel extends ConstructorReadable[LightGBMRegressionModel] {
+  def loadNativeModelFromFile(session: SparkSession, filename: String, labelColName: String = "label",
+                              featuresColName: String = "features",
+                              predictionColName: String = "prediction"): LightGBMRegressionModel = {
+    val uid = Identifiable.randomUID("LightGBMRegressor")
+    val textRdd = session.read.text(filename)
```
Comment (mhamilton723) [inline src/lightgbm/src/main/scala/LightGBMRegressor.scala]: Could you use SparkSession.getOrCreate() for a cleaner api here?

File: src/lightgbm/src/main/scala/LightGBMClassifier.scala [hunk 5]
```diff
@@ -130,4 +130,28 @@ class LightGBMClassificationModel(
...
 
-object LightGBMClassificationModel extends ConstructorReadable[LightGBMClassificationModel]
+object LightGBMClassificationModel extends ConstructorReadable[LightGBMClassificationModel] {
+  def loadNativeModelFromFile(session: SparkSession, filename: String, labelColName: String = "label",
+                              featuresColName: String = "features", predictionColName: String = "prediction",
+                              probColName: String = "probability",
+                              rawPredictionColName: String = "rawPrediction"): LightGBMClassificationModel = {
+    val uid = Identifiable.randomUID("LightGBMClassifier")
+    val actualNumClasses = 2
+    val textRdd = session.read.text(filename)
```
Comment (mhamilton723) [inline src/lightgbm/src/main/scala/LightGBMClassifier.scala]: likewise here

File: src/lightgbm/src/main/python/LightGBMRegressor.py [hunk 6]
```diff
@@ -31,8 +32,33 @@ def saveNativeModel(self, sparkSession, filename):
...
+                                                    featuresColName, predictionColName)
+        return JavaParams._from_java(java_model)
+
+    @staticmethod
+    def loadNativeModelFromString(sparkSession, model, labelColName="label",
+                                  featuresColName="features", predictionColName="prediction"):
+        """
+        Load the model from a native LightGBM model string.
+        """
+        ctx = sparkSession.sparkContext
```
Comment (mhamilton723) [inline src/lightgbm/src/main/python/LightGBMRegressor.py]: likewise there is a python equivalent to SparkSession.getOrCreate. Its something like getActiveSession

File: src/lightgbm/src/main/python/LightGBMRegressor.py [hunk 7]
```diff
@@ -31,8 +32,33 @@ def saveNativeModel(self, sparkSession, filename):
         jsession = sql_ctx.sparkSession._jsparkSession
         self._java_obj.saveNativeModel(jsession, filename)
 
+    @staticmethod
+    def loadNativeModelFromFile(sparkSession, filename, labelColName="label",
+                                featuresColName="features", predictionColName="prediction"):
+        """
+        Load the model from a native LightGBM text file.
+        """
+        jsession = sparkSession._jsparkSession
```
Comment (mhamilton723) [inline src/lightgbm/src/main/python/LightGBMRegressor.py]: likewise there is a python equivalent to SparkSession.getOrCreate. Its something like getActiveSession

File: src/lightgbm/src/main/python/LightGBMClassifier.py [hunk 8]
```diff
@@ -26,13 +27,40 @@ def saveNativeModel(self, sparkSession, filename):
...
+        return JavaParams._from_java(java_model)
+
+    @staticmethod
+    def loadNativeModelFromString(sparkSession, model, labelColName="label",
+                                  featuresColName="features", predictionColName="prediction",
+                                  probColName="probability", rawPredictionColName="rawPrediction"):
+        """
+        Load the model from a native LightGBM model string.
+        """
+        ctx = sparkSession.sparkContext
```
Comment (mhamilton723) [inline src/lightgbm/src/main/python/LightGBMClassifier.py]: likewise there is a python equivalent to SparkSession.getOrCreate. Its something like getActiveSession

File: src/lightgbm/src/main/python/LightGBMClassifier.py [hunk 9]
```diff
@@ -26,13 +27,40 @@ def saveNativeModel(self, sparkSession, filename):
...
         self._java_obj.saveNativeModel(jsession, filename)
 
+    @staticmethod
+    def loadNativeModelFromFile(sparkSession, filename, labelColName="label",
+                                featuresColName="features", predictionColName="prediction",
+                                probColName="probability", rawPredictionColName="rawPrediction"):
+        """
+        Load the model from a native LightGBM text file.
+        """
+        jsession = sparkSession._jsparkSession
```
Comment (mhamilton723) [inline src/lightgbm/src/main/python/LightGBMClassifier.py]: likewise there is a python equivalent to SparkSession.getOrCreate. Its something like getActiveSession

File: src/lightgbm/src/main/python/LightGBMClassifier.py [hunk 10]
```diff
@@ -26,13 +27,40 @@ def saveNativeModel(self, sparkSession, filename):
         """
         Save the booster as string format to a local or WASB remote location.
         """
-        ctx = SparkContext.getOrCreate()
-        sql_ctx = SQLContext.getOrCreate(ctx)
-        jsession = sql_ctx.sparkSession._jsparkSession
+        jsession = sparkSession._jsparkSession
```
Comment (mhamilton723) [inline src/lightgbm/src/main/python/LightGBMClassifier.py]: likewise there is a python equivalent to SparkSession.getOrCreate. Its something like getActiveSession

File: src/lightgbm/src/test/scala/VerifyLightGBMClassifier.scala (hunk 3 above)
Comment (imatiach-msft) [reply src/lightgbm/src/test/scala/VerifyLightGBMClassifier.scala]: done!

File: src/lightgbm/src/main/scala/LightGBMRegressor.scala (hunk 4 above)
Comment (imatiach-msft) [reply src/lightgbm/src/main/scala/LightGBMRegressor.scala]: done!

File: src/lightgbm/src/main/scala/LightGBMClassifier.scala (hunk 5 above)
Comment (imatiach-msft) [reply src/lightgbm/src/main/scala/LightGBMClassifier.scala]: done!

File: src/lightgbm/src/main/python/LightGBMRegressor.py (hunk 6 above)
Comment (imatiach-msft) [reply src/lightgbm/src/main/python/LightGBMRegressor.py]: done!

File: src/lightgbm/src/main/python/LightGBMRegressor.py (hunk 7 above)
Comment (imatiach-msft) [reply src/lightgbm/src/main/python/LightGBMRegressor.py]: done!

File: src/lightgbm/src/main/python/LightGBMClassifier.py (hunk 8 above)
Comment (imatiach-msft) [reply src/lightgbm/src/main/python/LightGBMClassifier.py]: done!

File: src/lightgbm/src/main/python/LightGBMClassifier.py (hunk 9 above)
Comment (imatiach-msft) [reply src/lightgbm/src/main/python/LightGBMClassifier.py]: done!

File: src/lightgbm/src/main/python/LightGBMClassifier.py (hunk 10 above)
Comment (imatiach-msft) [reply src/lightgbm/src/main/python/LightGBMClassifier.py]: done!

**Pull request merged**